src/Hosting/TestHost/src/AsyncStreamWrapper.cs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.TestHost
+{
+    internal class AsyncStreamWrapper : Stream
+    {
+        private Stream _inner;
+        private Func<bool> _allowSynchronousIO;
+
+        internal AsyncStreamWrapper(Stream inner, Func<bool> allowSynchronousIO)
```

Comment:
Is the reason for the Func so the value is checked on every operation between which the value could be changed?

Reply:
Right, the per request feature could be changed after the stream was created.